README.md
py-cpuinfo
==========

[![Latest Version](https://img.shields.io/pypi/v/py-cpuinfo.svg)](https://pypi.org/project/py-cpuinfo/)
[![License](https://img.shields.io/pypi/l/py-cpuinfo.svg)](https://github.com/workhorsy/py-cpuinfo/blob/master/LICENSE)
[![Python Version](https://img.shields.io/pypi/pyversions/py-cpuinfo.svg)](https://pypi.org/project/py-cpuinfo/)
[![PyPI - Downloads](https://img.shields.io/pypi/dm/py-cpuinfo)](https://pypi.org/project/py-cpuinfo/)

Py-cpuinfo gets CPU info with pure Python. Py-cpuinfo should work without any
extra programs or libraries, beyond what your OS provides. It does not require
any compilation(C/C++, assembly, et cetera) to use. It works with Python 3.


Example
-----
~~~python
if __name__ == '__main__':
    from cpuinfo import get_cpu_info

    for key, value in get_cpu_info().items():
        print("{0}: {1}".format(key, value))
~~~


API
-----
~~~python
'''
Returns the CPU info by using the best sources of information for your OS.
Returns the result in a dict
'''
get_cpu_info()

'''
Returns the CPU info by using the best sources of information for your OS.
Returns the result in a json string
'''
get_cpu_info_json()
~~~


Fields
-----
| key                           | Example value             | Return Format         |
| :---------------------------- | :------------------------ | :-------------------- |
| "python_version"              | "3.10.4.final.0 (64 bit)" | string                |
| "cpuinfo_version"             | (8, 0, 0)                 | (int, int, int)       |
| "cpuinfo_version_string"      | "8.0.0"                   | string                |
| "hz_advertised_friendly"      | "2.9300 GHz"              | string                |
| "hz_actual_friendly"          | "1.7330 GHz"              | string                |
| "hz_advertised"               | (2930000000, 0)           | (int, int)            |
| "hz_actual"                   | (1733000000, 0)           | (int, int)            |
| "arch"                        | "X86_64"                  | 'ARM_7', 'ARM_8', 'LOONG_32', 'LOONG_64', 'MIPS_32', 'MIPS_64', 'PPC_32', 'PPC_64', 'RISCV_32', 'RISCV_64', 'SPARC_32', 'SPARC_64', 'S390X', 'X86_32', 'X86_64'|
| "bits"                        | 64                        | int                   |
| "count"                       | 4                         | int                   |
| "l1_data_cache_size"          | 32768                     | int                   |
| "l1_instruction_cache_size"   | 32768                     | int                   |
| "l2_cache_size"               | 262144                    | int                   |
| "l2_cache_line_size"          | 256                       | int                   |
| "l2_cache_associativity"      | 6                         | int                   |
| "l3_cache_size"               | 3145728                   | int                   |
| "stepping"                    | 5                         | int                   |
| "model"                       | 30                        | int                   |
| "family"                      | 6                         | int                   |
| "processor_type"              | 0                         | int                   |
| "flags"                       | ['acpi', 'aperfmperf', 'apic', 'arch_perfmon', 'bts', 'clflush', 'cmov', 'constant_tsc', 'cx16', 'cx8', 'de', 'ds_cpl', 'dtes64', 'dtherm', 'dts', 'ept', 'est', 'flexpriority', 'fpu', 'fxsr', 'ht', 'ida', 'lahf_lm', 'lm', 'mca', 'mce', 'mmx', 'monitor', 'msr', 'mtrr', 'nonstop_tsc', 'nopl', 'nx', 'pae', 'pat', 'pbe', 'pdcm', 'pebs', 'pge', 'pni', 'popcnt', 'pse', 'pse36', 'rdtscp', 'rep_good', 'sep', 'smx', 'ss', 'sse', 'sse2', 'sse4_1', 'sse4_2', 'ssse3', 'syscall', 'tm', 'tm2', 'tpr_shadow', 'tsc', 'vme', 'vmx', 'vnmi', 'vpid', 'xtopology', 'xtpr'] | [string] |


Raw Fields
-----

These fields are pulled directly from the CPU and are unverified. They may
contain expected results. Other times they may contain wildly unexpected
results or garbage. So it would be a bad idea to rely on them.

| key                           | Example value             | Return Format         |
| :---------------------------- | :------------------------ | :-------------------- |
| "vendor_id_raw"               | "GenuineIntel"            | string                |
| "hardware_raw"                | "BCM2708"                 | string                |
| "brand_raw"                   | "Intel(R) Core(TM) i7 CPU         870  @ 2.93GHz" | string                |
| "arch_string_raw"             | "x86_64"                  | string                |



Run as a script
-----
~~~bash
$ python cpuinfo/cpuinfo.py
~~~

Run as a module
-----
~~~bash
$ python -m cpuinfo
~~~

Run as a library
-----
~~~python
if __name__ == '__main__':
    from cpuinfo import get_cpu_info
    info = get_cpu_info()
    print(info)
~~~

Run under Pyinstaller
-----
~~~python
# NOTE: Pyinstaller may spawn infinite processes if __main__ is not used
if __name__ == '__main__':
    from cpuinfo import get_cpu_info
    from multiprocessing import freeze_support

    # NOTE: Pyinstaller also requires freeze_support
    freeze_support()
    info = get_cpu_info()
    print(info)
~~~

Install instructions
-----
~~~bash
$ python -m pip install -U py-cpuinfo
~~~


Command Line Arguments
----
~~~
--help: show this help message and exit
--json: Return the info in JSON format
--version: Return the version of py-cpuinfo
--trace: Traces code paths used to find CPU info to file
~~~


OS Support
-----
| OS            | Tested and should work                                 | Untested        |
| :------------ | :----------------------------------------------------- | :-------------- |
| Android       |                                                        | Everything      |
| BSD           | FreeBSD, PC-BSD, TrueOS                                | OpenBSD, NetBSD |
| Cygwin        | Windows                                                |                 |
| Haiku         | Haiku Nightly                                          | BeOS            |
| Linux         | Arch, Centos, Debian, Fedora, Gentoo, OpenSuse, Ubuntu |                 |
| OS X          | 10.8 - 10.14                                           |                 |
| Solaris       | Oracle Solaris, OpenIndiana                            |                 |
| Windows       | XP, Vista, 7, 8, 10                                    | RT              |


CPU Support
-----
* X86 32bit and 64bit
* Some ARM, LoongArch, MIPS, PPC, RISCV, SPARC, and S390X CPUs


These approaches are used for getting info:
-----
1. Windows Registry (Windows)
2. /proc/cpuinfo (Linux)
3. sysctl (OS X)
4. dmesg (Unix/Linux)
5. /var/run/dmesg.boot (BSD/Unix)
6. isainfo and kstat (Solaris)
7. cpufreq-info (BeagleBone)

ChangeLog
* Release 9.1.0
	* Fixed Bug #186: Move to Python 3 style classes and default UTF-8 encoding
	* Fixed Bug #194: Add support for LoongArch

10/25/2022 Release 9.0.0
	* Fixed Bug #178: Changes to lscpu breaks parsing of cache info
	* Fixed Bug #152: CPU stepping, model, and family values are blank if 0
	* Fixed Bug #177: Officially drop support for Python 2
	* Fixed Bug #171: Replace Python 3.11 deprecated unittest.makeSuite
	* Fixed Bug #173: Fix lgtm.com alerts
	* Fixed Bug #165: Support Wheel
	* Fixed Bug #167: Add support for RISC-V

04/14/2021 Release 8.0.0
	* Fixed Bug #161: Accept arm64 as an alias for aarch64
	* Fixed Bug #160: Add MIPS architecture support
	* Fixed Bug #74: Add option to trace code paths to file
	* Fixed Bug #157: Remove multiple checks for sestatus

07/05/2020 Release 7.0.0
	* Fixed Bug #133: CPU flags vary between runs on Mac OS X
	* Fixed Bug #150: Change 'byte code' to 'machine code'
	* Fixed Bug #128: Overhead from generating machine code throws off CPUID HZ
	* Fixed Bug #136: On non BeOS systems, calling sysinfo may open GUI program
	* Fixed Bug #138: Invalid escape sequences warn when building in Python 3.8
	* Fixed Bug #147: Remove extended_model and extended_family fields
	* Fixed Bug #146: CPUID family and model is wrong
	* Fixed Bug #144: Cache fields should be full ints instead of kb strings

06/11/2020 Release 6.0.0
	* Fixed Bug #140: The get_cache function has swapped fields
	* Fixed Bug #142: Remove empty and zeroed fields
	* Fixed Bug #115: Missing data on Ryzen CPUs
	* Fixed Bug #122: Rename fields to be more clear
	* Fixed Bug #125: Add option to return --version
	* Fixed Bug #126: Make test suite also check SELinux
	* Fixed Bug #120: Make unit tests also test CPUID
	* Fixed Bug #69: Add s390x support

03/20/2019 Release 5.0.0
	* Fixed Bug #117: Remove PyInstaller hacks
	* Fixed Bug #108: Client script runs multiple times without __main__
	* Fixed Bug #113: Add option to return results in json
	* Fixed Bug #110: Always tries to run wmic in get_system_info.py

04/01/2018 Release 4.0.0
	* Fixed Bug #80: Broken when using Pyinstaller
	* Fixed Bug #77: Get L1, L2, and L3 cache info from lscpu
	* Fixed Bug #79: Byte formats are inconsistent
	* Fixed Bug #81: Byte formatter breaks on non strings
	* Fixed Bug #84: Fixed wrong setup of CPUID machine code and added more flags
	* Fixed Bug #88: Get Windows CPU info with wmic
	* Fixed Bug #94: Fails to detect winreg imported as _winreg in get_system_info
	* Fixed Bug #95: Include if Python is 32 or 64 bit in get_system_info
	* Fixed Bug #87: lscpu gets brand field twice
	* Fixed Bug #96: Include Python version in output
	* Fixed Bug #85: CPUID HZ measurement is scaled wrong
	* Fixed Bug #100: Officially drop support for Python 2.6
	* Fixed Bug #101: Made it only check the dmesg boot log on Linux
	* Fixed Bug #102: Parsing Haiku sysinfo fails

06/10/2017 Release 3.3.0
	* Fixed Bug #70: subprocess.Popen breaks under Windows GUI app
	* Fixed Bug #72: Fails to parse invalid CPUID result.

04/19/2017 Release 3.2.0
	* Fixed Bug #68: Broken Qemu guest CPU flag parsing on Ubuntu 17.04 POWER8

04/19/2017 Release 3.1.0
	* Fixed Bug #63: Include py-cpuinfo version in output
	* Fixed Bug #29: Missing AVX2 CPU flag on OS X
	* Fixed Bug #41: Broken on Odroid XU3 armhf ARM 32bit
	* Fixed Bug #61: Not working on RHEL7.3 ppc64le

tools/get_system_info.py
#!/usr/bin/env python

# Copyright (c) 2014-2022 Matthew Brennan Jones <matthew.brennan.jones@gmail.com>
# Py-cpuinfo gets CPU info with pure Python
# It uses the MIT License
# It is hosted at: https://github.com/workhorsy/py-cpuinfo
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


import os, sys
import glob
import platform
import multiprocessing
import subprocess

try:
	import _winreg as winreg
except ImportError as err:
	try:
		import winreg
	except ImportError as err:
		pass

is_windows = platform.system().lower() == 'windows'

out_file_name = 'system_info.txt'
out_file = open(out_file_name, 'w')


def run_and_get_stdout(command, pipe_command=None):
	from subprocess import Popen, PIPE

	if not pipe_command:
		p1 = Popen(command, stdout=PIPE, stderr=PIPE, stdin=PIPE)
		output = p1.communicate()[0]
		output = output.decode(encoding='UTF-8')
		return p1.returncode, output
	else:
		p1 = Popen(command, stdout=PIPE, stderr=PIPE, stdin=PIPE)
		p2 = Popen(pipe_command, stdin=p1.stdout, stdout=PIPE, stderr=PIPE)
		p1.stdout.close()
		output = p2.communicate()[0]
		output = output.decode(encoding='UTF-8')
		return p2.returncode, output

def program_paths(program_name):
	paths = []
	exts = filter(None, os.environ.get('PATHEXT', '').split(os.pathsep))
	for p in os.environ['PATH'].split(os.pathsep):
		p = os.path.join(p, program_name)
		if os.access(p, os.X_OK):
			paths.append(p)
		for e in exts:
			pext = p + e
			if os.access(pext, os.X_OK):
				paths.append(pext)
	return paths

def has_sestatus():
	return len(program_paths('sestatus')) > 0

def sestatus_b():
	return run_and_get_stdout(['sestatus', '-b'])

def _is_selinux_enforcing():
	# Just return if the SE Linux Status Tool is not installed
	if not has_sestatus():
		return False

	# Run the sestatus, and just return if it failed to run
	returncode, output = sestatus_b()
	if returncode != 0:
		return False

	# Figure out if explicitly in enforcing mode
	for line in output.splitlines():
		line = line.strip().lower()
		if line.startswith("current mode:"):
			if line.endswith("enforcing"):
				return True
			else:
				return False

	# Figure out if we can execute heap and execute memory
	can_selinux_exec_heap = False
	can_selinux_exec_memory = False
	for line in output.splitlines():
		line = line.strip().lower()
		if line.startswith("allow_execheap") and line.endswith("on"):
			can_selinux_exec_heap = True
		elif line.startswith("allow_execmem") and line.endswith("on"):
			can_selinux_exec_memory = True

	return (not can_selinux_exec_heap or not can_selinux_exec_memory)

def parse_arch(arch_string_raw):
	import re

	arch, bits = None, None
	arch_string_raw = arch_string_raw.lower()

	# X86
	if re.match(r'^i\d86$|^x86$|^x86_32$|^i86pc$|^ia32$|^ia-32$|^bepc$', arch_string_raw):
		arch = 'X86_32'
		bits = 32
	elif re.match(r'^x64$|^x86_64$|^x86_64t$|^i686-64$|^amd64$|^ia64$|^ia-64$', arch_string_raw):
		arch = 'X86_64'
		bits = 64
	# ARM
	elif re.match(r'^armv8-a|aarch64$', arch_string_raw):
		arch = 'ARM_8'
		bits = 64
	elif re.match(r'^armv7$|^armv7[a-z]$|^armv7-[a-z]$|^armv6[a-z]$', arch_string_raw):
		arch = 'ARM_7'
		bits = 32
	elif re.match(r'^armv8$|^armv8[a-z]$|^armv8-[a-z]$', arch_string_raw):
		arch = 'ARM_8'
		bits = 32
	# PPC
	elif re.match(r'^ppc32$|^prep$|^pmac$|^powermac$', arch_string_raw):
		arch = 'PPC_32'
		bits = 32
	elif re.match(r'^powerpc$|^ppc64$|^ppc64le$', arch_string_raw):
		arch = 'PPC_64'
		bits = 64
	# SPARC
	elif re.match(r'^sparc32$|^sparc$', arch_string_raw):
		arch = 'SPARC_32'
		bits = 32
	elif re.match(r'^sparc64$|^sun4u$|^sun4v$', arch_string_raw):
		arch = 'SPARC_64'
		bits = 64
	# S390X
	elif re.match(r'^s390x$', arch_string_raw):
		arch = 'S390X'
		bits = 64
	# RISCV
	elif re.match(r'^riscv$|^riscv32$|^riscv32be$', arch_string_raw):
		arch = 'RISCV_32'
		bits = 32
	elif re.match(r'^riscv64$|^riscv64be$', arch_string_raw):
		arch = 'RISCV_64'
		bits = 64

	return (arch, bits)

def print_output(name, output):
	line = "=" * 79
	out_file.write('{0}:\n{1}\n{2}\n\n\n\n'.format(name, line, output))

print_output('sys.executable', sys.executable)

print_output('sys.version_info', sys.version_info)

if hasattr(sys, 'maxsize'):
	sizes = {
		2**31-1: '32 bit',
		2**63-1: '64 bit',
	}
	friendly_maxsize = sizes.get(sys.maxsize) or 'unknown'
	print_output('sys.maxsize', "{0} ({1})".format(sys.maxsize, friendly_maxsize))

print_output('multiprocessing.cpu_count', multiprocessing.cpu_count())

if hasattr(os, 'cpu_count'):
	print_output('os.cpu_count', os.cpu_count())

if 'NUMBER_OF_PROCESSORS' in os.environ:
	print_output("os.environ['NUMBER_OF_PROCESSORS']", os.environ['NUMBER_OF_PROCESSORS'])

if hasattr(os, 'sysconf'):
	print_output("os.sysconf('SC_NPROCESSORS_ONLN')", os.sysconf('SC_NPROCESSORS_ONLN'))

if program_paths('sysctl'):
	returncode, output = run_and_get_stdout(['sysctl', '-n', 'hw.ncpu'])
	print_output('sysctl -n hw.ncpu', output)

print_output('platform.uname', platform.uname())

print_output('platform.architecture', platform.architecture())

print_output('platform.system', platform.system())

print_output('platform.machine', platform.machine())


if program_paths('cpufreq-info'):
	returncode, output = run_and_get_stdout(['cpufreq-info'])
	print_output('cpufreq-info', output)

if program_paths('sestatus'):
	returncode, output = run_and_get_stdout(['sestatus', '-b'])
	print_output('sestatus -b', output)

if os.path.exists('/proc/cpuinfo'):
	returncode, output = run_and_get_stdout(['cat', '/proc/cpuinfo'])
	print_output('cat /proc/cpuinfo', output)

if program_paths('sysctl'):
	returncode, output = run_and_get_stdout(['sysctl', 'machdep.cpu', 'hw.cpufrequency'])
	print_output('sysctl machdep.cpu hw.cpufrequency', output)

if program_paths('sysctl'):
	returncode, output = run_and_get_stdout(['sysctl', 'hw.model', 'hw.machine'])
	print_output('sysctl hw.model hw.machine', output)

if program_paths('isainfo'):
	returncode, output = run_and_get_stdout(['isainfo', '-vb'])
	print_output('isainfo -vb', output)

if program_paths('kstat'):
	returncode, output = run_and_get_stdout(['kstat', '-m', 'cpu_info'])
	print_output('kstat -m cpu_info', output)

if program_paths('lscpu'):
	returncode, output = run_and_get_stdout(['lscpu'])
	print_output('lscpu', output)

if program_paths('dmesg'):
	returncode, output = run_and_get_stdout(['dmesg', '-a'])
	if returncode != 0:
		returncode, output = run_and_get_stdout(['dmesg'])
	if len(output) > 20480:
		output = output[0 : 20480]
	print_output('dmesg -a', output)

if os.path.exists('/var/run/dmesg.boot'):
	returncode, output = run_and_get_stdout(['cat', '/var/run/dmesg.boot'])
	if len(output) > 20480:
		output = output[0 : 20480]
	print_output('cat /var/run/dmesg.boot', output)

if program_paths('sysinfo'):
	uname = platform.system().strip().strip('"').strip("'").strip().lower()
	is_beos = 'beos' in uname or 'haiku' in uname
	if is_beos:
		returncode, output = run_and_get_stdout(['sysinfo', '-cpu'])
		print_output('sysinfo -cpu', output)

if program_paths('lsprop'):
	ibm_features = glob.glob('/proc/device-tree/cpus/*/ibm,pa-features')
	if ibm_features:
		returncode, output = run_and_get_stdout(['lsprop', ibm_features[0]])
		print_output('lsprop /proc/device-tree/cpus/*/ibm,pa-features', output)

if program_paths('wmic'):
	returncode, output = run_and_get_stdout(['wmic', 'os', 'get', 'Version'])
	if returncode == 0 and len(output) > 0:
		returncode, output = run_and_get_stdout(['wmic', 'cpu', 'get', 'Name,CurrentClockSpeed,L2CacheSize,L3CacheSize,Description,Caption,Manufacturer', '/format:list'])
		if returncode == 0 and len(output) > 0:
			print_output('wmic cpu get Name,CurrentClockSpeed,L2CacheSize,L3CacheSize,Description,Caption,Manufacturer /format:list', output)

if 'winreg' in sys.modules or '_winreg' in sys.modules:
	key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"Hardware\Description\System\CentralProcessor\0")
	processor_brand = winreg.QueryValueEx(key, "ProcessorNameString")[0]
	winreg.CloseKey(key)
	print_output('winreg processor_brand', processor_brand)

	key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"Hardware\Description\System\CentralProcessor\0")
	vendor_id = winreg.QueryValueEx(key, "VendorIdentifier")[0]
	winreg.CloseKey(key)
	print_output('winreg vendor_id', vendor_id)

	key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"SYSTEM\CurrentControlSet\Control\Session Manager\Environment")
	arch_string_raw = winreg.QueryValueEx(key, "PROCESSOR_ARCHITECTURE")[0]
	winreg.CloseKey(key)
	print_output('winreg arch_string_raw', arch_string_raw)

	key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"Hardware\Description\System\CentralProcessor\0")
	hz_actual = winreg.QueryValueEx(key, "~Mhz")[0]
	winreg.CloseKey(key)
	print_output('winreg hz_actual', hz_actual)

	key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"Hardware\Description\System\CentralProcessor\0")
	feature_bits = winreg.QueryValueEx(key, "FeatureSet")[0]
	winreg.CloseKey(key)
	print_output('winreg feature_bits', feature_bits)


class ASM:
	def __init__(self, restype=None, argtypes=(), machine_code=None):
		self.restype = restype
		self.argtypes = argtypes
		self.machine_code = machine_code
		self.prochandle = None
		self.mm = None
		self.func = None
		self.address = None
		self.size = 0
		self.is_selinux_enforcing = _is_selinux_enforcing()

	def compile(self):
		import ctypes

		machine_code = bytes.join(b'', self.machine_code)
		self.size = ctypes.c_size_t(len(machine_code))

		if is_windows:
			# Allocate a memory segment the size of the machine code, and make it executable
			size = len(machine_code)
			# Alloc at least 1 page to ensure we own all pages that we want to change protection on
			if size < 0x1000: size = 0x1000
			MEM_COMMIT = ctypes.c_ulong(0x1000)
			PAGE_READWRITE = ctypes.c_ulong(0x4)
			pfnVirtualAlloc = ctypes.windll.kernel32.VirtualAlloc
			pfnVirtualAlloc.restype = ctypes.c_void_p
			self.address = pfnVirtualAlloc(None, ctypes.c_size_t(size), MEM_COMMIT, PAGE_READWRITE)
			if not self.address:
				raise Exception("Failed to VirtualAlloc")

			# Copy the machine code into the memory segment
			memmove = ctypes.CFUNCTYPE(ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t)(ctypes._memmove_addr)
			if memmove(self.address, machine_code, size) < 0:
				raise Exception("Failed to memmove")

			# Enable execute permissions
			PAGE_EXECUTE = ctypes.c_ulong(0x10)
			old_protect = ctypes.c_ulong(0)
			pfnVirtualProtect = ctypes.windll.kernel32.VirtualProtect
			res = pfnVirtualProtect(ctypes.c_void_p(self.address), ctypes.c_size_t(size), PAGE_EXECUTE, ctypes.byref(old_protect))
			if not res:
				raise Exception("Failed VirtualProtect")

			# Flush Instruction Cache
			# First, get process Handle
			if not self.prochandle:
				pfnGetCurrentProcess = ctypes.windll.kernel32.GetCurrentProcess
				pfnGetCurrentProcess.restype = ctypes.c_void_p
				self.prochandle = ctypes.c_void_p(pfnGetCurrentProcess())
			# Actually flush cache
			res = ctypes.windll.kernel32.FlushInstructionCache(self.prochandle, ctypes.c_void_p(self.address), ctypes.c_size_t(size))
			if not res:
				raise Exception("Failed FlushInstructionCache")
		else:
			from mmap import mmap, MAP_PRIVATE, MAP_ANONYMOUS, PROT_WRITE, PROT_READ, PROT_EXEC

			# Allocate a private and executable memory segment the size of the machine code
			machine_code = bytes.join(b'', self.machine_code)
			self.size = len(machine_code)
			self.mm = mmap(-1, self.size, flags=MAP_PRIVATE | MAP_ANONYMOUS, prot=PROT_WRITE | PROT_READ | PROT_EXEC)

			# Copy the machine code into the memory segment
			self.mm.write(machine_code)
			self.address = ctypes.addressof(ctypes.c_int.from_buffer(self.mm))

		# Cast the memory segment into a function
		functype = ctypes.CFUNCTYPE(self.restype, *self.argtypes)
		self.func = functype(self.address)

	def run(self):
		# Call the machine code like a function
		retval = self.func()

		return retval

	def free(self):
		import ctypes

		# Free the function memory segment
		if is_windows:
			MEM_RELEASE = ctypes.c_ulong(0x8000)
			ctypes.windll.kernel32.VirtualFree(ctypes.c_void_p(self.address), ctypes.c_size_t(0), MEM_RELEASE)
		else:
			self.mm.close()

		self.prochandle = None
		self.mm = None
		self.func = None
		self.address = None
		self.size = 0

class CPUID:
	def __init__(self):
		# Figure out if SE Linux is on and in enforcing mode
		self.is_selinux_enforcing = _is_selinux_enforcing()

	def _asm_func(self, restype=None, argtypes=(), machine_code=None):
		asm = ASM(restype, argtypes, machine_code)
		asm.compile()
		return asm

	def _run_asm(self, *machine_code):
		import ctypes

		asm = ASM(ctypes.c_uint32, (), machine_code)
		asm.compile()
		retval = asm.run()
		asm.free()
		return retval

	# http://en.wikipedia.org/wiki/CPUID#EAX.3D0:_Get_vendor_ID
	def get_vendor_id(self):
		rets = []

		# EBX
		ebx = self._run_asm(
			b"\x31\xC0",        # xor eax,eax
			b"\x0F\xA2"         # cpuid
			b"\x89\xD8"         # mov ax,bx
			b"\xC3"             # ret
		)
		#print('!!! ebx: ', hex(ebx))
		rets.append(ebx)

		# ECX
		ecx = self._run_asm(
			b"\x31\xC0",        # xor eax,eax
			b"\x0f\xa2"         # cpuid
			b"\x89\xC8"         # mov ax,cx
			b"\xC3"             # ret
		)
		#print('!!! ecx: ', hex(ecx))
		rets.append(ecx)

		# EDX
		edx = self._run_asm(
			b"\x31\xC0",        # xor eax,eax
			b"\x0f\xa2"         # cpuid
			b"\x89\xD0"         # mov ax,dx
			b"\xC3"             # ret
		)
		#print('!!! edx: ', hex(edx))
		rets.append(edx)

		return rets

	# http://en.wikipedia.org/wiki/CPUID#EAX.3D1:_Processor_Info_and_Feature_Bits
	def get_info(self):
		# EAX
		eax = self._run_asm(
			b"\xB8\x01\x00\x00\x00",   # mov eax,0x1"
			b"\x0f\xa2"                # cpuid
			b"\xC3"                    # ret
		)
		#print('!!! eax: ', hex(eax))
		return eax

	# http://en.wikipedia.org/wiki/CPUID#EAX.3D80000000h:_Get_Highest_Extended_Function_Supported
	def get_max_extension_support(self):
		# Check for extension support
		max_extension_support = self._run_asm(
			b"\xB8\x00\x00\x00\x80" # mov ax,0x80000000
			b"\x0f\xa2"             # cpuid
			b"\xC3"                 # ret
		)

		return max_extension_support

	# http://en.wikipedia.org/wiki/CPUID#EAX.3D1:_Processor_Info_and_Feature_Bits
	def get_flags(self, max_extension_support):
		rets = []

		# EDX
		edx = self._run_asm(
			b"\xB8\x01\x00\x00\x00",   # mov eax,0x1"
			b"\x0f\xa2"                # cpuid
			b"\x89\xD0"                # mov ax,dx
			b"\xC3"                    # ret
		)
		#print('!!! edx: ', hex(edx))
		rets.append(edx)

		# ECX
		ecx = self._run_asm(
			b"\xB8\x01\x00\x00\x00",   # mov eax,0x1"
			b"\x0f\xa2"                # cpuid
			b"\x89\xC8"                # mov ax,cx
			b"\xC3"                    # ret
		)
		#print('!!! ecx: ', hex(ecx))
		rets.append(ecx)

		# http://en.wikipedia.org/wiki/CPUID#EAX.3D7.2C_ECX.3D0:_Extended_Features
		if max_extension_support >= 7:
			# EBX
			ebx = self._run_asm(
				b"\x31\xC9",            # xor ecx,ecx
				b"\xB8\x07\x00\x00\x00" # mov eax,7
				b"\x0f\xa2"         # cpuid
				b"\x89\xD8"         # mov ax,bx
				b"\xC3"             # ret
			)
			#print('!!! ebx: ', hex(ebx))
			rets.append(ebx)

			# ECX
			ecx = self._run_asm(
				b"\x31\xC9",            # xor ecx,ecx
				b"\xB8\x07\x00\x00\x00" # mov eax,7
				b"\x0f\xa2"         # cpuid
				b"\x89\xC8"         # mov ax,cx
				b"\xC3"             # ret
			)
			#print('!!! ecx: ', hex(ecx))
			rets.append(ecx)

		# http://en.wikipedia.org/wiki/CPUID#EAX.3D80000001h:_Extended_Processor_Info_and_Feature_Bits
		if max_extension_support >= 0x80000001:
			# EBX
			ebx = self._run_asm(
				b"\xB8\x01\x00\x00\x80" # mov ax,0x80000001
				b"\x0f\xa2"         # cpuid
				b"\x89\xD8"         # mov ax,bx
				b"\xC3"             # ret
			)
			#print('!!! ebx: ', hex(ebx))
			rets.append(ebx)

			# ECX
			ecx = self._run_asm(
				b"\xB8\x01\x00\x00\x80" # mov ax,0x80000001
				b"\x0f\xa2"         # cpuid
				b"\x89\xC8"         # mov ax,cx
				b"\xC3"             # ret
			)
			#print('!!! ecx: ', hex(ecx))
			rets.append(ecx)

		return rets


	# http://en.wikipedia.org/wiki/CPUID#EAX.3D80000002h.2C80000003h.2C80000004h:_Processor_Brand_String
	def get_processor_brand(self, max_extension_support):
		rets = []

		# Processor brand string
		if max_extension_support >= 0x80000004:
			instructions = [
				b"\xB8\x02\x00\x00\x80", # mov ax,0x80000002
				b"\xB8\x03\x00\x00\x80", # mov ax,0x80000003
				b"\xB8\x04\x00\x00\x80"  # mov ax,0x80000004
			]
			for instruction in instructions:
				# EAX
				eax = self._run_asm(
					instruction,  # mov ax,0x8000000?
					b"\x0f\xa2"   # cpuid
					b"\x89\xC0"   # mov ax,ax
					b"\xC3"       # ret
				)
				#print('!!! eax: ', hex(eax))
				rets.append(eax)

				# EBX
				ebx = self._run_asm(
					instruction,  # mov ax,0x8000000?
					b"\x0f\xa2"   # cpuid
					b"\x89\xD8"   # mov ax,bx
					b"\xC3"       # ret
				)
				#print('!!! ebx: ', hex(ebx))
				rets.append(ebx)

				# ECX
				ecx = self._run_asm(
					instruction,  # mov ax,0x8000000?
					b"\x0f\xa2"   # cpuid
					b"\x89\xC8"   # mov ax,cx
					b"\xC3"       # ret
				)
				#print('!!! ecx: ', hex(ecx))
				rets.append(ecx)

				# EDX
				edx = self._run_asm(
					instruction,  # mov ax,0x8000000?
					b"\x0f\xa2"   # cpuid
					b"\x89\xD0"   # mov ax,dx
					b"\xC3"       # ret
				)
				#print('!!! edx: ', hex(edx))
				rets.append(edx)

		return rets

	# http://en.wikipedia.org/wiki/CPUID#EAX.3D80000006h:_Extended_L2_Cache_Features
	def get_cache(self, max_extension_support):
		# Just return if the cache feature is not supported
		if max_extension_support < 0x80000006:
			return

		# ECX
		ecx = self._run_asm(
			b"\xB8\x06\x00\x00\x80"  # mov ax,0x80000006
			b"\x0f\xa2"              # cpuid
			b"\x89\xC8"              # mov ax,cx
			b"\xC3"                   # ret
		)

		return ecx

def get_cpu_info_from_cpuid():
	# Just return if not X86
	arch_string_raw = platform.machine()
	arch, bits = parse_arch(arch_string_raw)
	if not arch in ['X86_32', 'X86_64']:
		return {}

	# FIXME: Return none if SE Linux is in enforcing mode
	cpuid = CPUID()
	if cpuid.is_selinux_enforcing:
		return {}

	output = ''
	max_extension_support = cpuid.get_max_extension_support()
	output += '# max_extension_support\n'
	output += hex(max_extension_support) + ',\n'

	rets = cpuid.get_cache(max_extension_support)
	output += '# get_cache\n'
	output += hex(rets) + ',\n'

	rets = cpuid.get_info()
	output += '# get_info\n'
	output += hex(rets) + ',\n'

	rets = cpuid.get_processor_brand(max_extension_support)
	output += '# get_processor_brand\n'
	output += ', '.join([hex(n) for n in rets]) + ',\n'

	rets = cpuid.get_vendor_id()
	output += '# get_vendor_id\n'
	output += ', '.join([hex(n) for n in rets]) + ',\n'

	rets = cpuid.get_flags(max_extension_support)
	output += '# get_flags\n'
	output += ', '.join([hex(n) for n in rets]) + ',\n'

	return output


try:
	output = get_cpu_info_from_cpuid()
	print_output('CPUID', output)
except Exception:
	pass

out_file.close()
print('System info written to "{0}"'.format(out_file_name))
